test/model/recent_senders_test.dart

```diff
@@ -0,0 +1,229 @@
+}
+
+void main() {
+  group('RecentSenders', () {
+    test('starts with empty stream and topic senders', () {
+      final model = RecentSenders();
+      checkMatchesMessages(model, []);
+    });
+
+    test('stream message gets included, DM message gets ignored', () {
+      final model = RecentSenders();
+      final streamMessage = eg.streamMessage();
+      final dmMessage = eg.dmMessage(from: eg.selfUser, to: [eg.otherUser]);
+
+      model.handleMessage(streamMessage);
+      model.handleMessage(dmMessage);
+      checkMatchesMessages(model, [streamMessage]);
+    });
+
+    group('add a batch of messages', () {
+      late RecentSenders model;
+      final stream1 = eg.stream(streamId: 1);
+      final user10 = eg.user(userId: 10);
+
+      void setupModel(List<Message> messages) {
+        model = RecentSenders();
+        for (final message in messages) {
+          model.handleMessage(message);
+        }
+      }
+
+      test('batch goes before the existing messages', () {
+        final existingMessages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 300),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 400),
+        ];
+        setupModel(existingMessages);
+
+        final messages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 100),
+          eg.streamMessage(stream: stream1, topic: 'b', sender: user10, id: 200),
+        ];
+        model.handleMessages(messages);
+
+        checkMatchesMessages(model,
+          [...existingMessages, ...messages]..sort((m1, m2) => m1.id.compareTo(m2.id)));
+      });
+
+      test('batch goes after the existing messages', () {
+        final existingMessages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 300),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 400),
+        ];
+        setupModel(existingMessages);
+
+        final messages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 500),
+          eg.streamMessage(stream: stream1, topic: 'b', sender: user10, id: 600),
+        ];
+        model.handleMessages(messages);
+
+        checkMatchesMessages(model,
+          [...existingMessages, ...messages]..sort((m1, m2) => m1.id.compareTo(m2.id)));
+      });
+
+      test('batch is interspersed among the existing messages', () {
+        final existingMessages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 200),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 400),
+        ];
+        setupModel(existingMessages);
+
+        final messages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 100),
+          eg.streamMessage(stream: stream1, topic: 'b', sender: user10, id: 300),
+          eg.streamMessage(stream: stream1, topic: 'b', sender: user10, id: 500),
+        ];
```

Comment:
Does this do what the test name says it does? It looks like it doesn't:

stream 1, topic 'a' was [200, 400], and gets 100, which goes at the beginning
stream 1, topic 'b' was absent and gets some messages

Interspersed would look like: some tracker had [100, 300, 500], and gets [200, 400].

Reply:
Oh, I guess they are interspersed on the stream tracker.

Still, I think the test is easiest to read and think about if there's just one tracker involved of each kind. So changing the two `topic: 'b'` to `topic: 'a'` would do it.

Reply:
My bad. Those messages were supposed to be in `topic: 'a'`, but after copy-pasting it from the previous revision, I forgot to rename it.
And the new messages are supposed to scatter around and between the old ones.